### PR TITLE
Dynamic solver loading

### LINKED ISSUE #4: Shift to better SAT solving
Could consider embedding a pre-compiled SAT solver that is wrapped similar to DSHARP. [Cadical](https://github.com/arminbiere/cadical) fits the bill for that.

### PR DESCRIPTION
**Pull request by haz**
Closes #4

### CODE REVIEW

**Comment by haz** [comment]
@blyxxyz Can you have a look at the mypy issue with this? Problem is that `is_CNF` doesn't indicate the typing at all to mypy, and I'm not sure how best to approach it (this is the first project I've seen it used).

**Comment by blyxxyz** [comment]
It might be possible to indicate with an [overload](https://docs.python.org/3/library/typing.html#typing.overload) that `is_CNF()` returns `False` if its argument isn't `And[Or[Var]]`, but I couldn't figure it out.
I think the safety isn't worth the trouble here. In that case it's ok to add a `# type: ignore` comment to suppress the error or use a [cast](https://docs.python.org/3/library/typing.html#typing.cast). See 0750552e74b2651fc65759b6bf9c537e98d47e29.

**Comment by haz** [comment]
This cast was precisely the thing I was looking for (stumbling around in the dark with asserts).

Think there should be a test conditioned on system compatibility that runs `kissat` through the same paces as the native solver?

**Comment by blyxxyz** [comment]
`nnf.kissat` should also be added to `docs/nnf.rst` to generate documentation, see 290ba998883d5fc00704e27c0fcc7b4877f73134. I forgot about that for #6.

File: nnf/__init__.py [hunk 1]
```diff
@@ -596,6 +598,18 @@ def to_CNF(self) -> 'And[Or[Var]]':
         return tseitin.to_CNF(self)
 
     def _cnf_satisfiable(self) -> bool:
+        """Call a SAT solver on the presumed CNF theory."""
+        if 'native' == SAT_BACKEND:
```
[the remark this answers is not in the document]
Comment (haz) [reply nnf/__init__.py:602]: Old habits die hard ;)

File: nnf/kissat.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    infd, infname = tempfile.mkstemp(text=True)
+    try:
+        with open(infd, 'w') as f:
+            dimacs.dump(sentence, f, mode='cnf', var_labels=var_labels_inverse)
+
+        proc = subprocess.Popen(
+            args + [infname],
+            stdout=subprocess.PIPE,
+            universal_newlines=True
+        )
```
[the remark this answers is not in the document]
Comment (haz) [reply nnf/kissat.py:49]: Appears to be the same on windows.

File: MANIFEST.in [hunk 3]
```diff
@@ -0,0 +1 @@
+recursive-include nnf/bin *
```
[the remark this answers is not in the document]
Comment (haz) [reply MANIFEST.in:1]: Nice catch! 1.2M is entirely fine.

File: nnf/bin/README.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Binary files included with python-nnf
+
+## kissat
+
+The `kissat` SAT solver is released under MIT license, and can be found [here](https://github.com/arminbiere/kissat/). The binary included in this directory corresponds to the statically linked version at [this commit](https://github.com/arminbiere/kissat/commit/baef4609163f542dc08f43aef02ce8da0581a2b5). No modifications were made to the `kissat` solver.
```
Comment (blyxxyz) [inline nnf/bin/README.md:5]: I think we'd have to include the license itself too, e.g. as a file `LICENSE.kissat`, based on ["The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software."](https://github.com/arminbiere/kissat/blob/master/LICENSE)

**Comment by haz** [comment]
How's it look now?

Not sure if I responded to your `solve()` question, but absolutely it makes sense. Most of the functionality is already there.

**Pull request merged**